### ISSUE TITLE
Sorting point of orders by point of order category

### ISSUE THREAD

**Issue by r-peschke** [OpenSlides/openslides-backend#1774]
- new collection point_of_order_category. The category-list will be used by all list_of_speakers of the meeting. Columns
  - id: int 
  - text:string
  - rank: int
  - meeting_id: RelationField
  - If made per relation a RelationListField  speaker_ids is necessary
- collection meeting with new columns
  -  point_of_order_category_ids as RelationListField
  - point_of_order_category_enabled: bool
- collection speaker with new column point_of_order_category_id **or** text and rank from point_of_order_category
- Action speaker.create: a new optional field for point_of_order_category. Putting point of order on top is default, the client should know what to do (manual or sorting by poo-category)   
- Action speaker.sort: Writes the received list speaker_ids with a new weight to database, no changes I suppose

**Comment by reiterl** [comment]
Should the development of this feature happen in an own feature branch?

**Comment by jsangmeister** [comment]
> Decision taken: The collection speaker don't get a column for point_of_order_category_id and the opposite relation

I don't like this decision, it's unnecessary data redundancy and might lead to confusing situations when e.g. a point of order text gets changed, but the existing speakers don't update and keep the old text.

> because for sorting and showing in the client we only need the rank and the category text

Which can be directly accessed via the relation, so this is an argument for neither solution

> One more advantage, if needed: The sorting order keeps clear, even if the ranking will be changed afterwards.

The sorting order in the client is only determined by the `weight`. Currently, if a point of order is added, all subsequent speakers' weights get incremented to make room for the new speaker. I would assume that the same method could and would be used for the new system, the the sorting order will not change anyway when the rank of a point of order changes.

**Comment by luisa-beerboom** [comment]
Where will the point_of_order_category_ids be set? How will the models be created?
I.e.: What action will control the point_of_order_categories?

**Comment by r-peschke** [comment]
> Where will the point_of_order_category_ids be set? How will the models be created? I.e.: What action will control the point_of_order_categories? Will it be meeting update, or will here be separate point_of_order_category actions available?

There will be the usual (create, update, delete) actions for the point of order category

**Comment by r-peschke** [comment]
> > Decision taken: The collection speaker don't get a column for point_of_order_category_id and the opposite relation
> 
> I don't like this decision, it's unnecessary data redundancy and might lead to confusing situations when e.g. a point of order text gets changed, but the existing speakers don't update and keep the old text.
> 

Will use the id of point_of_order_category

> > because for sorting and showing in the client we only need the rank and the category text
> 
> Which can be directly accessed via the relation, so this is an argument for neither solution
> 
> > One more advantage, if needed: The sorting order keeps clear, even if the ranking will be changed afterwards.
> 
> The sorting order in the client is only determined by the `weight`. Currently, if a point of order is added, all subsequent speakers' weights get incremented to make room for the new speaker. I would assume that the same method could and would be used for the new system, the the sorting order will not change anyway when the rank of a point of order changes.

There are no plans to change anything in this part for the client. For adding a new point-of_order the speaker-instance need, also existing behavior, the information, what kind of speaker it is. This issue only adds the new information of poo-category.

**Comment by reiterl** [comment]
This issue shouldn't be developed in a feature branch.

**Comment by reiterl** [comment]
Add the action descriptions to the wiki. @r-peschke : can you please look at them?
I will continue with the model and the category actions.

**Comment by luisa-beerboom** [comment]
I think it would make sense for the `text` property in `point_of_order` to be unique per meeting

**Comment by r-peschke** [comment]
> I think it would make sense for the `text` property in `point_of_order` to be unique per meeting

Yes, but I hope  that the user will be capable to manage this in about 20-30 categories

**Comment by luisa-beerboom** [comment]
It would make things safer on my end

**Comment by luisa-beerboom** [comment]
As far as I can see there is currently no defined way for the client to set `point_of_order_category_enabled`. It is not mentioned in the meeting.update wiki entry, nor in any other one for all I can tell.

@r-peschke Could you add the specification for this?

**Comment by jsangmeister** [comment]
Good catch, added it to https://github.com/OpenSlides/OpenSlides/wiki/meeting.update.

Two remarks:
- The `restriction_mode` is currently `C`, while all other settings have restriction mode `A`. I would suggest to change the restriction mode to `A` to match the other settings.
- The new setting breaks multiple naming patterns, e.g. all settings are prefixed by their collection/settings group, in this case it would be `list_of_speakers`. To match other settings like `list_of_speakers_enable_point_of_order_speakers` even more, I would suggest to rename the new setting to `list_of_speakers_enable_point_of_order_categories`.

I already included these changes in https://github.com/OpenSlides/OpenSlides/wiki/meeting.update. Let me know what you think.

**Comment by jsangmeister** [comment]
Talked with @r-peschke and we will implemented the changes as described. @reiterl you will need to update this PR accordingly and @luisa-beerboom you will need to update the client. If it's urgent, I could also implement the changes now.

**Comment by jsangmeister** [comment]
@r-peschke we both confused the _permission group_ as described in https://github.com/OpenSlides/OpenSlides/wiki/meeting.update with the _restriction mode_ in the `models.yml` :D Nevertheless, most other settings have restriction mode `B` instead of `C`, so we should follow that. I updated my earlier comment accordingly.

**Comment by luisa-beerboom** [comment]
> Talked with @r-peschke and we will implemented the changes as described. @reiterl you will need to update this PR accordingly and @luisa-beerboom you will need to update the client. If it's urgent, I could also implement the changes now.

From a development perspective, it would be very helpful if it worked as soon as possible, whether it is actually _urgent_ or not, depends on whether or not @rrenkert needs a working prototype to be ready at the end of this week

**Comment by luisa-beerboom** [comment]
> Good catch, added it to https://github.com/OpenSlides/OpenSlides/wiki/meeting.update.
> 
> Two remarks:
> 
>     * The `restriction_mode` is currently `C`, while all other settings have restriction mode `B` (see https://github.com/OpenSlides/openslides-backend/blob/ff5ccfc64dc0c55cdc18d5e65d1ba44445978421/global/meta/models.yml#L1020
>        and following). I would suggest to change the restriction mode to `B` to match the other settings.
> 
>     * The new setting breaks multiple naming patterns, e.g. all settings are prefixed by their collection/settings group, in this case it would be `list_of_speakers`. To match other settings like `list_of_speakers_enable_point_of_order_speakers` even more, I would suggest to rename the new setting to `list_of_speakers_enable_point_of_order_categories`.
> 
> 
> I already included these changes in https://github.com/OpenSlides/OpenSlides/wiki/meeting.update. Let me know what you think.

Will you also be using restriction mode `B` for the meetings `point_of_order_category_ids` relation?

**Comment by jsangmeister** [comment]
> Will you also be using restriction mode `B` for the meetings `point_of_order_category_ids` relation?

I was not planning to, do you want that changed?

**Comment by luisa-beerboom** [comment]
> > Will you also be using restriction mode `B` for the meetings `point_of_order_category_ids` relation?
> 
> I was not planning to, do you want that changed?

Yes, since that relation will be presented as a setting in the client

**Comment by jsangmeister** [comment]
Ok
